src/elexmodel/models/BaseElectionModel.py

```diff
@@ -264,3 +266,37 @@ def get_coefficients(self):
         These coefficients are for the point prediciton only, not for the lower or upper intervals models.
         """
         return self.features_to_coefficients
+
+    def find_optimal_lambda(
+        self,
+        reporting_units,
+        possible_lambda_values=[],
+        estimand="",
+        K=3,
+    ):
+        if len(possible_lambda_values) == 0:
+            return 0, None
+        if len(possible_lambda_values) == 1:
+            return possible_lambda_values[0], None
+
+        MAPE_arr = np.zeros_like(possible_lambda_values)  # array of MAPE sums for each lambda input
+
+        # loop through each lambda
+        for loc, lam in enumerate(possible_lambda_values):
+            for train_index, test_index in math_utils.get_kfold_splits(reporting_units, K):
+                # build model with custom lambda
+                train = reporting_units.iloc[train_index]
+                test = reporting_units.iloc[test_index]
+                unit_predictions = self.get_unit_predictions(train, test, estimand, lam)
+                MAPE = math_utils.compute_error(
+                    test[f"results_{estimand}"].values, unit_predictions.values, type_="mape"
+                )
+                MAPE_arr[loc] += MAPE
+
+        # determine average and best
+        MAPE_arr_avg = MAPE_arr / len(possible_lambda_values)
```

Comment:
I think the average should be diving by K, since that is how many runs we did for each possible lambda value

Reply:
Thanks for catching that! Changed the value- the assertions changed as well (the best average_MAPE is now ~0.83 rather than 0.625), but since we are just dividing by a different number, the MAPE calculations I did by hand are still consistent and the optimal lambda for both tests are still the same

Reply:
Randomizing the order of reporting units resulted in, often, more than one lambda value having the same average_MAPE regardless of dataset size. The smaller the dataset, the more unique the lambda values. Further, the average MAPE changes consistently so I had to remove the assertions for calculations done by hand.